File: src/backtracking/python/k_colorable_configurations/src/k_colorable_configurations.py
from typing import List, Dict


class Graph:
    """A simple undirected graph represented using an adjacency dictionary."""

    def __init__(self) -> None:
        """Initialize an empty graph."""
        self.num_vertices: int = 0
        self.adjacency_dict: Dict[int, List[int]] = {}

    def add_edge(self, u: int, v: int) -> None:
        """
        Add an undirected edge between vertices u and v.

        Args:
            u: First vertex.
            v: Second vertex.
        """
        if u not in self.adjacency_dict:
            self.adjacency_dict[u] = []

        if v not in self.adjacency_dict:
            self.adjacency_dict[v] = []

        self.adjacency_dict[u].append(v)
        self.adjacency_dict[v].append(u)
        if len(self.adjacency_dict) > self.num_vertices:
            self.num_vertices = len(self.adjacency_dict)


def find_k_colorable_configurations(G: Graph, k: int) -> List[List[int]]:
    """
    Find all valid k-colorable configurations of the graph.

    A k-colorable configuration assigns one of k colors to each vertex
    such that no two adjacent vertices share the same color.

    Args:
        G: The graph to color.
        k: The number of available colors.

    Returns:
        List of all valid configurations, where each configuration is a list
        of color assignments for each vertex.
    """

    def dfs(
        G: Graph, v: int, colors: List[int], configuration: List[int]
    ) -> None:
        if v == G.num_vertices:
            configurations.append(configuration.copy())
            return

        for color in range(k):
            if is_colorable(G, v, color, colors):
                colors[v] = color
                configuration.append(color)
                dfs(G, v + 1, colors, configuration)
                configuration.pop()
                colors[v] = -1

    def is_colorable(G: Graph, v: int, color: int, colors: List[int]) -> bool:
        """Check if assigning color to vertex v is valid."""
        for u in G.adjacency_dict[v]:
            if colors[u] == color:
                return False
        return True

    configurations: List[List[int]] = []
    dfs(G, 0, [-1] * G.num_vertices, [])
    return configurations

File: src/backtracking/cpp/k_colorable_configurations/src/k_colorable_configurations.cpp
#include "k_colorable_configurations.h"
#include <functional>

Graph::Graph() { num_vertices = 0; }

void Graph::add_edge(int u, int v) {
  if (adjacency_dict.count(u) == 0) {
    adjacency_dict[u] = {};
  }

  if (adjacency_dict.count(v) == 0) {
    adjacency_dict[v] = {};
  }

  adjacency_dict[u].push_back(v);
  adjacency_dict[v].push_back(u);
  if (adjacency_dict.size() > num_vertices) {
    num_vertices = adjacency_dict.size();
  }
}

std::vector<std::vector<int>> find_k_colorable_configurations(Graph &G, int k) {
  std::vector<std::vector<int>> configurations;

  std::function<bool(Graph &, int, int, std::vector<int> &)> is_colorable;
  is_colorable = [&](Graph &G, int v, int color, std::vector<int> &colors) {
    for (int u : G.adjacency_dict[v]) {
      if (colors[u] == color) {
        return false;
      }
    }
    return true;
  };

  std::function<void(Graph &, int, std::vector<int> &, std::vector<int> &)> dfs;
  dfs = [&](Graph &G, int v, std::vector<int> &colors,
            std::vector<int> &configuration) {
    if (v == G.num_vertices) {
      configurations.push_back(configuration);
      return;
    }

    for (int color = 0; color < k; ++color) {
      if (is_colorable(G, v, color, colors)) {
        colors[v] = color;
        configuration.push_back(color);
        dfs(G, v + 1, colors, configuration);
        configuration.pop_back();
        colors[v] = -1;
      }
    }
  };

  auto colors = std::vector<int>(G.num_vertices, -1);
  auto configuration = std::vector<int>();
  dfs(G, 0, colors, configuration);
  return configurations;
}